### DISCUSSION TITLE
Get a launch number in junit5 test

### DISCUSSION THREAD

**Discussion by mourav2222** [reportportal/reportportal#2052] [Q&A] [answered]
Hello,

i want to set a new attribute Launch with current Launch number in reportportal.

![image](https://github.com/reportportal/reportportal/assets/76436807/f8387c9a-1c34-4065-a0dc-a3edd7ceed4d)

If i get the obejct with current launch uuid, i don't found the function to get a lauch number , it is equal in report portal. E.g. 303.
```java
ReportPortalExtension.REPORT_PORTAL.withLaunch(Maybe.just("1688462634550:5b01c66e-e653-4fa3-8fa5-2f7ca5bb96b0"))
```

Can you say me, how i can get this number and set a new attribute "Launch" with current launch number in the ReportPortal of the running launch.

thx

**Comment by DzmitryHumianiuk** [comment, marked as answer]
@mourav2222 
I'll do my best to answer based on what I think you're asking. 
If this isn't what you're looking for, please rephrase your question.

My answer is based on the assumption that you want to obtain the launch number (`#303` in this specific example) to add it as the launch attribute during JUnit execution.

### Step 1: 
#### Get the launch `ID` (database record) and `NUMBER`  based on the `UUID` (unique identifier at the moment of the async reporting).
Request
```
GET ​/v1​/{projectName}​/launch​/uuid​/{launchId}
```

Response (example)
```json
{
  "owner": "superadmin",
📍"id": 282674,
  "uuid": "33765c80-b483-4a0e-bc3f-d69743d1d418",
  "name": "RP Smoke Test",
📍"number": 87,
  "startTime": 1692115442959,
  "endTime": 1692115532966,
  "lastModified": 1692115533446,
  "status": "FAILED",
  "statistics": {
    "executions": {
      "total": 190,
      "passed": 158,
      "failed": 32
    },
    "defects": {
      "to_investigate": {
        "total": 32,
        "ti001": 32
      }
    }
  },
  "attributes": [],
  "mode": "DEFAULT",
  "analysing": [],
  "approximateDuration": 104.5388,
  "hasRetries": false,
  "rerun": false,
  "metadata": {
    "qualityGate": {
      "id": 2,
      "status": "FAILED"
    },
    "rp.cluster.lastRun": "1692115538035"
  }
}
```

### Step 2: 
#### PUT extra attributes into launch by launch ID (database record).
Request
```
PUT
​/v1​/{projectName}​/launch​/{launchId}​/update
```
```bash
curl -X PUT "https://{SERVER}/api/v1/{projectName}/launch/{launchId}​/update" 
   -H  "accept: */*" 
   -H  "Content-Type: application/json" 
   -H  "Authorization: bearer token" 
   -d "{
     "attributes": [
       {
         "key": "Launch",
         "value": "87"
       }
     ]
   }"
```



Response 
```json
{
  "message": "Launch with ID = '282674' successfully updated."
}
```

<img width="440" alt="image" src="https://github.com/reportportal/reportportal/assets/11332788/6aba470c-fe21-498b-a0e5-e720cb9aefb5">

**Comment by mourav2222** [comment]
Hello Dzmitry,

i want to clear up my problem, perhaps you have some idea, how to solve a problem?
i have 2 widgets, but the left widget haven't launch number like a right widget. I must use in the left widget the attribute like now 'build' attribute from a demo. I think to save a launch number  in the my attribute (LauchNumber=#303) and afterwards to use it instead the 'build' attribute. My web ui tests run with junit5 and i use an extension @ExtendWith(ReportPortalExtension.class). Can i get a launch number in the wigdet oder i have to create a custom attribute with launch number like so (LauchNumber=#303) . If 2. option is valid, how can i do this in the junit5? I have not uuid of the started launch.

![Screenshot from 2023-08-22 20-15-59](https://github.com/reportportal/reportportal/assets/76436807/b6378283-81ad-47b6-b9f9-f955e33d84e2)

best regards

PS: My Solution

[Test worker] INFO qa.guru.allure.MyReportPortalExtension - Launch Number: https://github.com/reportportal/reportportal/issues/6
---------------------------------
```java
package qa.guru.allure;

import com.epam.reportportal.junit5.ReportPortalExtension;
import com.epam.reportportal.service.Launch;
import com.epam.ta.reportportal.ws.model.launch.LaunchResource;
import io.reactivex.Maybe;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.epam.reportportal.junit5.ItemType.SUITE;

public class MyReportPortalExtension extends ReportPortalExtension {

    private static final Logger LOGGER = LoggerFactory.getLogger(MyReportPortalExtension.class);

    @Override
    public void beforeAll(ExtensionContext context) {
//        super.beforeAll(context);

        Launch launch = getLaunch(context); // Trigger launch start
        String launchUuid = launch.getLaunch().blockingGet();
        LOGGER.info("Launch UUID: {}", launchUuid);

        Maybe<LaunchResource> launchResourceMaybe = launch.getClient().getLaunchByUuid(launch.getLaunch().blockingGet());
        Long launchNumber = launchResourceMaybe.blockingGet().getNumber();
        LOGGER.info("Launch Number: #{}", launchNumber);

        startTestItem(context, SUITE);
    }
}
```
```bash
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by retrofit2.Platform (file:/home/mike/.gradle/caches/modules-2/files-2.1/com.squareup.retrofit2/retrofit/2.9.0/d8fdfbd5da952141a665a403348b74538efc05ff/retrofit-2.9.0.jar) to constructor java.lang.invoke.MethodHandles$Lookup(java.lang.Class,int)
WARNING: Please consider reporting this to the maintainers of retrofit2.Platform
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[Test worker] INFO com.epam.reportportal.service.Launch - Rerun: false
[Test worker] INFO qa.guru.allure.MyReportPortalExtension - Launch UUID: 332fdb0f-e77d-43fc-85e2-a634954787fe
[Test worker] INFO qa.guru.allure.MyReportPortalExtension - Launch Number: #6
idea: true
[Test worker] INFO com.codeborne.selenide.impl.WebDriverThreadLocalContainer - No webdriver is bound to current thread: 1 - let's create a new webdriver
[Test worker] INFO com.codeborne.selenide.impl.FileHelper - Creating folder: /home/mike/IdeaProjects/allure-github-example/build/downloads/1692733823401_1865907_1
[Test worker] INFO com.codeborne.selenide.webdriver.SelenideNettyClientFactory - Changed readTimeout from PT3M to PT1M30S
[Test worker] INFO com.codeborne.selenide.webdriver.SelenideNettyClientFactory - Changed readTimeout from PT3M to PT1M30S
Aug 22, 2023 9:50:26 PM org.openqa.selenium.devtools.CdpVersionFinder findNearestMatch
WARNING: Unable to find an exact match for CDP version 116, so returning the closest version found: 115
[Test worker] INFO com.codeborne.selenide.webdriver.WebDriverFactory - BrowserName=chrome Version=116.0.5845.96 Platform=linux
```
![image](https://github.com/reportportal/reportportal/assets/76436807/fca8f445-cd63-4a39-aaeb-1c6ec9ea7e50)

**Reply by DzmitryHumianiuk** [reply]
@HardNorth do we have Launch update method implemented in Java-client?

**Comment by mourav2222** [comment]
@DzmitryHumianiuk 

Now launch numbers are equal in the two widgets.  Thx for hints.

![image](https://github.com/reportportal/reportportal/assets/76436807/a2605291-ca55-4dba-a798-627a95a8f69f)

**Reply by DzmitryHumianiuk** [reply]
@mourav2222 glad you've being able to make it.
